src/libmusig2.c

```diff
@@ -235,196 +236,122 @@ int musig2_aggregate_pubkey(musig2_context *mc, secp256k1_pubkey *pk_list, int n
-    secp256k1_xonly_pubkey_serialize(mcs->mc->ctx, ser_pk, &xonly_pk);
-    musig2_key_agg_coef(mcs->mc, ser_pk, param.a, mcs->mc->L);
-
-    /* Get x_only version of aggregated public key and its parity */
-    assert(secp256k1_xonly_pubkey_from_pubkey(mcs->mc->ctx, &xonly_aggr_pk, &param.par_pk, &mcs->mc->aggr_pk));
-    secp256k1_xonly_pubkey_serialize(mcs->mc->ctx, param.ser_aggr_pk, &xonly_aggr_pk);
+    musig2_key_agg_coef(&mcs->mc, ser_pk, a);
 
+    assert(secp256k1_xonly_pubkey_from_pubkey(mcs->mc.ctx, &xonly_temp, NULL, &mcs->mc.aggr_pk));
+    secp256k1_xonly_pubkey_serialize(mcs->mc.ctx, ser_aggr_pk, &xonly_temp);
 
-    /* Compute `b`, `R`, and `c` */
-    param.par_R = 0;
-    musig2_calc_b(mcs, &param);
-    if (!musig2_calc_R(mcs, &param)){
+    musig2_calc_b(&mcs->mc, ser_aggr_pk, b, msg, msg_len);
+    /* Compute `R` */
+    if (!musig2_calc_R(&mcs->mc, b, b_LIST)){
         printf("Failed to calculate R. \n");
         return 0;
     }
-    musig2_calc_c(mcs->mc, &param);
 
-    /* If par_R is 1, then negate b, call Calc_R again, and compute parsig */
-    if (param.par_R == 1 && param.par_pk == 0){
-        if (!secp256k1_ec_seckey_negate(mcs->mc->ctx, param.b)){
-            printf("Failed to negate b. \n");
-            return 0;
-        }
-        if (!musig2_calc_R(mcs, &param)){
-            printf("Failed to calculate R. \n");
-            return 0;
-        }
-        param.par_R = -1;
-        if (!musig2_set_parsig(mcs, &param, mps->sig)){
-            printf("Failed to generate partial signature. \n");
-            return 0;
-        }
-    }
-    /* If par_pk is 1, negate c and compute parsig */
-    else if (param.par_R == 0 && param.par_pk == 1){
-        if (!secp256k1_ec_seckey_negate(mcs->mc->ctx, param.c)){
-            printf("Failed to negate c. \n");
-            return 0;
-        }
-        if (!musig2_set_parsig(mcs, &param, mps->sig)){
-            printf("Failed to generate partial signature. \n");
-            return 0;
-        }
-    }
-    /* If par_pk == par_R, compute parsig */
-    else{
-        if (!musig2_set_parsig(mcs, &param, mps->sig)){
-            printf("Failed to generate partial signature. \n");
-            return 0;
-        }
-    }
+    assert(secp256k1_xonly_pubkey_from_pubkey(mcs->mc.ctx, &xonly_temp, NULL, &mcs->mc.aggr_R));
+    secp256k1_xonly_pubkey_serialize(mcs->mc.ctx, ser_aggr_R, &xonly_temp);
+
+    memcpy(bytes_to_hash, &ser_aggr_R, XONLY_BYTES);
+    memcpy(&bytes_to_hash[XONLY_BYTES], &ser_aggr_pk, XONLY_BYTES);
+    memcpy(&bytes_to_hash[XONLY_BYTES * 2], msg, msg_len);
 
-    memcpy(mps->R.data, mcs->mc->aggr_R.data, PK_BYTES);
+    assert(secp256k1_tagged_sha256(mcs->mc.ctx, c, (const unsigned char *)"BIP0340/challenge" , 17, bytes_to_hash, sizeof (bytes_to_hash)));
 
-    /* Update the state after each signature */
-    mcs->mc->state++;
+    memcpy(mps->R.data, mcs->mc.aggr_R.data, PK_BYTES);
```

Comment:
We should remove `aggr_R` from the signer context and use `R` from the partial signature context. So, we no longer need to copy `R` to mps.